src/appleseed/renderer/modeling/bsdf/alsurfacelayerbrdf.cpp

```diff
@@ -0,0 +1,482 @@
+            Spectrum f_value;
+            const float layer_probability = fresnel_term(*values, wo, m, f_value);
+
+            const MDF<float>& mdf = pick_mdf(values->m_distribution);
+            float probability = layer_probability * reflection_pdf(*values, mdf, wo, m);
+
+            probability += values->m_osl_bsdf->evaluate_pdf(
+                values->m_closure_data,
+                geometric_normal,
+                shading_basis,
+                outgoing,
+                incoming,
+                modes) * (1.0f - layer_probability);
+
+            return probability;
+        }
+
+        float sample_ior(
+            SamplingContext&        sampling_context,
+            const void*             data) const APPLESEED_OVERRIDE
+        {
+            // Forward to substrate.
+            const InputValues* values = static_cast<const InputValues*>(data);
+            return values->m_osl_bsdf->sample_ior(
+                sampling_context,
+                values->m_closure_data);
+
+        }
+
+        void compute_absorption(
+            const void*             data,
+            const float             distance,
+            Spectrum&               absorption) const APPLESEED_OVERRIDE
+        {
+            // Forward to substrate.
+            const InputValues* values = static_cast<const InputValues*>(data);
+            values->m_osl_bsdf->compute_absorption(
+                values->m_closure_data,
+                distance,
+                absorption);
+        }
+
+      private:
+        typedef AlSurfaceLayerBRDFInputValues InputValues;
+
+        enum MicrofacetDistribution
+        {
+            Beckmann = 0,
+            GGX
+        };
+
+        enum FresnelMode
+        {
+            Dielectric = 0,
+            Metallic
+        };
+
+        static const MDF<float>& pick_mdf(int distribution)
+        {
+            if (distribution == Beckmann)
+                return m_beckmann_mdf;
+            else
+                return m_ggx_mdf;
+        }
+
+        static float fresnel_term(
+            const InputValues&  values,
+            Vector3f            wo,
+            const Vector3f&     m,
+            Spectrum&           value)
+        {
+            const Vector3f n(0.0f, 1.0f, 0.0f);
+
+            if (wo.y < 0.0f)
+                wo = -wo;
+
+            if (values.m_fresnel_mode == Dielectric)
+            {
+                const FresnelDielectricFun f(
+                    values.m_reflectance,
+                    1.0f,
+                    1.0f / values.m_ior);
+                f(wo, m, n, value);
+            }
+            else
+            {
+                const FresnelFriendlyConductorFun f(
+                    values.m_normal_reflectance,
+                    values.m_edge_tint,
+                    1.0f);
+                f(wo, m, n, value);
+                value *= values.m_reflectance;
+            }
+
+            return clamp(max_value(value), 0.0f, 1.0f);
```

Comment:
Value can be a color if fresnel mode is metallic or if the reflectance is a color.
What alSurface does (if I read the code correctly) is:

F * reflectance * microfacet(...) + (1.0 - max_value(F * reflectance)) * substrate()